models/optional_modules/consent/snowplow_web_cmp_stats.sql

```diff
@@ -0,0 +1,85 @@
+{{
+  config(
+    materialized='table',
+    sql_header=snowplow_utils.set_query_tag(var('snowplow__query_tag', 'snowplow_dbt'))
+  )
+}}
+
+with ordering as (
+
+  select
+    event_id,
+    domain_userid,
+    derived_tstamp,
+    event_name,
+    event_type,
+    lag(event_name, 1) over(partition by domain_userid order by derived_tstamp) as lag_type,
+    lag(event_id, 1) over(partition by domain_userid order by derived_tstamp) as lag_id
+
+  from {{ ref('snowplow_web_consent_log') }}
+
+  where event_type <> 'pending' or event_type is null
+
+)
+
+, cmp_grouping as (
+
+  select
+    event_id,
+    domain_userid,
+    derived_tstamp,
+    event_name,
+    case when event_name = 'cmp_visible' then event_id
+      when lag_type = 'cmp_visible' then lag_id
+      else lag_id end as cmp_id
+
+  from ordering
+
+)
+
+, event_orders as (
+
+  select
+    event_id,
+    domain_userid,
+    cmp_id,
+    derived_tstamp,
+    row_number() over(partition by cmp_id order by derived_tstamp) as row_num
+
+  from cmp_grouping
+
+)
+
+, first_consent_events as (
+
+  select
+    l.event_id,
+    l.domain_userid,
+    l.page_view_id,
+    l.domain_sessionid,
+    l.elapsed_time as cmp_visible_time,
+    l.derived_tstamp as cmp_tstamp,
+    -- without this it would contain duplicates for those that have followup events
+    max(case when o.derived_tstamp = l.derived_tstamp then null
+      else o.derived_tstamp end) as first_consent_event_tstamp
+
+  from {{ ref('snowplow_web_consent_log') }} l
+
+  left join event_orders o
+
+  on o.cmp_id = l.event_id
+
+  where l.event_name = 'cmp_visible' and (o.row_num = 2 or o.derived_tstamp = l.derived_tstamp)
```

Comment:
We are gathering all the cmp_visible events and the next consent_event after that (if there is any). Above you will find the overall concept in my previous comment.

Reply:
Makes sense, I hadn't quite registered the join wasn't on event_id on both sides. Just for my own curiosity - what happens if the user gets the cmp_visible then just leaves the site, do they just not have a first consent event? So their records just don't get in the output table?

Reply:
They should still receive a 'pending' consent event. Not sure if it will be enforced or rather just illustrated as the recommended tracking action but I will follow this logic for modelling regardless.